crates/gitbutler-branch-actions/src/branch.rs

```diff
@@ -0,0 +1,395 @@
+                    "Failed to process branch group {:?} to branch entry: {:?}",
+                    identity,
+                    branch_entry
+                );
+            }
+            branch_entry.ok()
+        })
+        .collect();
+    Ok(branches)
+}
+
+/// Converts a group of branches with the same identity into a single branch entry
+fn branch_group_to_branch(
+    identity: BString,
+    group_branches: Vec<&GroupBranch>,
+    repo: &git2::Repository,
+    local_author: &Author,
+) -> Result<BranchListing> {
+    let virtual_branch = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Virtual(vb) => Some(vb),
+            _ => None,
+        })
+        .next();
+    let remote_branches: Vec<&gix::Reference> = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Git(gb) => Some(gb),
+            _ => None,
+        })
+        .filter(|reference| {
+            reference.remote(gix::remote::Direction::Fetch).is_some()
+                || reference.remote(gix::remote::Direction::Push).is_some()
+        })
+        .filter(|reference| matches!(reference.target(), gix::refs::TargetRef::Peeled(_)))
+        .collect();
+    let local_branches: Vec<&gix::Reference> = group_branches
+        .iter()
+        .filter_map(|branch| match branch {
+            GroupBranch::Git(gb) => Some(gb),
+            _ => None,
+        })
+        .filter(|reference| {
+            reference.remote(gix::remote::Direction::Fetch).is_none()
+                && reference.remote(gix::remote::Direction::Push).is_none()
+        })
+        .filter(|reference| matches!(reference.target(), gix::refs::TargetRef::Peeled(_)))
+        .collect();
+
+    // Virtual branch associated with this branch
+    let virtual_branch_reference = virtual_branch.map(|branch| VirtualBranchReference {
+        given_name: branch.name.clone(),
+        id: branch.id,
+        in_workspace: branch.in_workspace,
+    });
+
+    let remotes: Vec<BString> = remote_branches
+        .iter()
+        .map(|reference| {
+            let fetch = reference.remote_name(gix::remote::Direction::Fetch);
+            let push = reference.remote_name(gix::remote::Direction::Push);
+            fetch.unwrap_or(push.unwrap()).as_bstr().into() // Already checked that at least one of them is present
+        })
+        .collect();
+
+    // The head commit for which we calculate statistics.
+    // If there is a virtual branch let's get it's head. Alternatively, pick the first local branch and use it's head.
+    // If there are no local branches, pick the first remote branch.
+    let head = if let Some(vbranch) = virtual_branch {
+        Some(vbranch.head)
+    } else if let Some(reference) = local_branches.first().cloned() {
+        Some(git2::Oid::from_bytes(reference.id().as_bytes())?) // We have filtered out symbolic references, which would otherwise panic
```

Comment:
@Byron this is the place where i had an issue, but now i am filtering on line 119 and 131